text/0569-tracked-list.md

```diff
@@ -0,0 +1,306 @@
+
+    return first;
+  }
+}
+```
+
+From a change tracking perspective, we can see that this is actually very
+similar to the immutable pattern described at the beginning of this RFC - we
+are essentially telling the tracking system that some changes have occured, and
+now `friends` is _effectively_ a new array that needs to be reprocessed.
+
+In general, methods that _access_ the list and don't mutate it will entangle
+the list in the autotracking context. Methods that _mutate_ the list will
+invalidate, but not entangle the list, preventing backflow issues when
+rendering.
+
+### `getAt` and `setAt`
+
+It will not be possible to access the list items directly using `[]` syntax,
+like on normal arrays, since this would require native proxies (and would incur
+performance penalties). `TrackedList` will have two extra methods instead for
+direct index access: `getAt` and `setAt`. These methods will operate similar to
+`objectAt` and `replace` on `EmberArray` today, allowing the user to pass a
+numeric index as the first value, and in the case of `setAt` to pass the value
+to set the index to as the second:
+
+```ts
+declare class TrackedList<T> {
+  getAt(index: number): T;
+  setAt(index: number, value: T): T;
+}
+```
+
+### `sort` and `reverse`
+
+`sort` and `reverse` are unusual array methods, because they both _read_ the
+value, _and_ mutate the array, in place. Most of the time, users want to do one
+or the other, and from a change tracking perspective this is not ideal behavior.
+
+As such, `TrackedList` will instead always return a new list when either of
+these methods are called, allowing users to derive state without accidentally
+mutating upstream values.
+
+### `arr` in Array Callbacks
+
+Many array callbacks, such as `map` and `reduce`, receive the array as the third
+argument:
+
+```js
+[1, 2, 3].map((value, index, arr) => value + 1);
+```
+
+For performance reasons, `TrackedList` will not provide the list itself as the
+last argument. Instead, it will provide the underlying array, which is used as a
+storage mechanism to back the list. In `DEBUG` mode, this array will be wrapped
+with a Proxy which will:
+
+1. Alert users if they attempt to use non-standard methods such as `getAt` and
+   `setAt`.
+2. Throw an error if users try to mutate the array in any way. It should be
+   considered read-only for the purposes of usage within callbacks.
+
+This will allow common array-mapping libraries to be used interoperably with
+lists, without causing issues by accidentally mutating the internal storage of
+the list.
+
+### `Array.isArray`, `isEmberArray`, and Instance Checks
+
+`Array.isArray` and `isEmberArray` will return `false` for instances of
+`TrackedList`. Users can check to see if it is an instance using `instanceOf`:
+
+```js
+let friends = new TrackedList();
+
+friends instanceOf TrackedList; // true
+```
+
+### `length`
+
+`length` will be an immutable/readonly property on `TrackedList`. This is to
```

Comment:
```suggestion
`length` will be a readonly property on `TrackedList`. This is to
```

I *think* if I'm understanding the intent here correctly, we should describe this as `readonly`, since it *will* change if users `push` or `pop` from the list.

Also, is `length` auto-tracked? (I'd assume yes, but I'd love if it were explicit here.)